### DISCUSSION TITLE
useQueries combine result not memoized

### DISCUSSION THREAD

**Discussion by oblador** [TanStack/query#8122] [General] [answered]
According to documentation, the way to reduce re-renders with `useQuery` is to use a referentially stable `select` option that will then be memoized. 

With `useQueries` there is no `select` option, but a `combine` which functionally is essentially the same. However even if it returns a referentially equal result it will still cause a re-render every time a dependent query is updated. This means there is no convenient way to optimise queries with many dependencies. IMHO it would be preferable if only instances where the combined result changed would cause rerenders.

Related: https://github.com/TanStack/query/pull/7233

Example/reproduction: https://stackblitz.com/edit/vite-react-ts-jnyua9?file=src%2FApp.tsx

**Comment by paulobmarcos** [comment]
I agree with this one, as it would make the behaviour similar to `useQuery` and would avoid unnecessary re-renders.

Probably having a `select` would be overkill, since `combine` can already combine only the things we want, but a `select` could still be useful.

Let's say we want to combine multiple results, but in different situations we would like to still reuse the same `useQueries` abstraction, as it handles all of the necessary calls, but "subscribe" to different pieces of information.

**Reply by paulobmarcos** [reply]
Brief Example:

```ts
const useMyData = (select) => {
   return useQueries({ ..., select });
}
```

```tsx
const ComponentA = () => {
  const result = useMyData((combinedData) => data.id);
}

const ComponentB = () => {
  const result = useMyData((combinedData) => data.name);
}
```

**Comment by TkDodo** [comment]
hmm yeah there is something wrong. if combine returns a stable value, there shouldn’t be any more re-renders 🤔

**Reply by oblador** [reply]
Want me to convert this into an issue?

**Reply by TkDodo** [reply]
sure, would you like to take a stab at fixing it?

**Reply by oblador** [reply]
I started looking into this and I'm unsure how to properly implement this to be honest. For example the options passed to QueryObserver are never used , neither are the ones passed to setQueries(). I'm lacking a lot of context here, but I think it would require a non-trivial refactor to actually make use of the combine option inside the observer or do something patchy in useQueries().

**Reply by TkDodo** [reply, marked as answer]
I have a fix here:
- https://github.com/TanStack/query/pull/8153

**Reply by oblador** [reply]
I can confirm this fixes it for me! Thank you so much for the quick fix 🥳

**Reply by TkDodo** [reply]
awesome 🎉

**Reply by mrudowski** [reply]
Hi, 

I think a have a problem with this change - please take a look on this example and click "Update first query result".
Component with useQueries doesn't refresh with new value.
https://stackblitz.com/edit/vite-react-ts-iybfwe?file=src%2FApp.tsx

in version 5.59.3 - before this change - it works fine
https://stackblitz.com/edit/vite-react-ts-vl8q1r?file=src%2FApp.tsx

**Reply by TkDodo** [reply]
Thanks for reporting @mrudowski. I debugged this a bit and found out that this is a problem in property tracking. Somewhere along the way, we are losing our property tracking, so what gets passed to `combine` is a plain object, and we think you’re only interested in the `pending` prop. Since that prop doesn’t change when you call `setQueryData`, the re-render gets skipped.

This should definitely work, I’ll have to look deeper where we are losing that. If you want a workaround, you need to access `.data` before you do the early return:

https://stackblitz.com/edit/vite-react-ts-ck3x9f

**Reply by TkDodo** [reply]
I think I got it:
- https://github.com/TanStack/query/pull/8295

**Reply by mrudowski** [reply]
You are the best!

I had another example here and it working fine too now in v5.60.5
https://stackblitz.com/edit/vite-react-ts-qpz8tj?file=src%2FApp.tsx

So big thanks for quick fix and for all your hard work
